File: src/python/problems/EraseOverlapIntervals.py
# https://leetcode.com/problems/non-overlapping-intervals/

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class EraseOverlapIntervals:
    def Solution(self, intervals: List[List[int]]) -> int:
        result = 0
        intervals.sort(key = lambda x: x[0])
        current = intervals[0]
        for i in range(1, len(intervals)):
            if current[1] > intervals[i][0]:
                result += 1
                current = current if current[1] <= intervals[i][1] else intervals[i]
            else:
                current = intervals[i]

        return result



EraseOverlapIntervals().Solution([[1,2],[2,3],[3,4],[1,3]])

File: src/csharp/Problems/EraseOverlapIntervals.cs
//https://leetcode.com/problems/non-overlapping-intervals/

namespace LeetCode.Problems;

public sealed class EraseOverlapIntervals : ProblemBase
{
    [Theory]
    [ClassData(typeof(EraseOverlapIntervals))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.Param2dArray("[[1,2],[2,3],[3,4],[1,3]]").Result(1))
          .Add(it => it.Param2dArray("[[1,2],[1,2],[1,2]]").Result(2))
          .Add(it => it.Param2dArray("[[1,2],[2,3]]").Result(0));

    private int Solution(int[][] intervals)
    {
        var result = 0;
        Array.Sort(intervals, (x, y) =>  x[0] - y[0]);

        var current = intervals[0];
        for (var i = 1; i < intervals.Length; i++)
        {
            if (current[1] > intervals[i][0])
            {
                result++;
                current = current[1] <= intervals[i][1] ? current : intervals[i];
            }
            else
            {
                current = intervals[i];
            }
        }

        return result;
    }
}